torch/csrc/jit/tensorexpr/mem_dependency_checker.cpp

```diff
@@ -0,0 +1,1261 @@
+    knownVarBounds_[pair.first] = pair.second;
+  }
+
+  if (currentScope_ != prev_scope) {
+    mergeScope(currentScope_, prev_scope, true);
+    currentScope_ = prev_scope;
+  }
+
+  // handle kernel outputs.
+  if (prev_scope->block == nullptr) {
+    insertBuffers(outputs_, AccessType::Output);
+  }
+}
+
+void MemDependencyChecker::visit(const Let* v) {
+  const Stmt* last = lastStmt_;
+  lastStmt_ = v;
+
+  IRVisitor::visit(v);
+
+  lastStmt_ = last;
+
+  const Var* var = v->var();
+  if (knownVarBounds_.count(var) != 0) {
+    currentScope_->shadowedVarBounds[var] = knownVarBounds_[var];
+  }
+
+  currentScope_->localVars.insert(var);
+  knownVarBounds_[var] = {v->value(), v->value()};
+}
+
+// Don't support AtomicAdd yet, it's a bit more complex since it's both a read
+// and a write. It's only inserted during Cuda codegen so this should be okay.
+void MemDependencyChecker::visit(const AtomicAdd* v) {
+  throw std::runtime_error("MemDependencyChecker AtomicAdd unimplemented");
+}
+
+void MemDependencyChecker::updateWriteHistory(
+    std::list<BoundRelationship>& writeHistory,
+    const std::shared_ptr<AccessInfo>& info,
+    int latestAccessToClose,
+    bool insert) {
+  bool isWrite = info->isWrite();
+
+  for (auto it = writeHistory.begin(); it != writeHistory.end();) {
+    auto& indexBounds = it->first;
+    std::shared_ptr<AccessInfo> other = it->second;
+    if (info->hasDependency(other)) {
+      ++it;
+      continue;
+    }
+
+    OverlapKind overlap = overlaps(indexBounds, info->bounds());
+
+    if (overlap == NoOverlap) {
+      ++it;
+      continue;
+    }
+
+    // Only writes can close open accesses.
+    if (!isWrite) {
+      info->addDependency(other);
+      other->addDependent(info);
+      ++it;
+      continue;
+    }
+
+    // If we're not closing accesses we can stop here.
+    if ((int)other->id() > latestAccessToClose) {
+      ++it;
+      continue;
+    }
+
+    if (overlap == TotalOverlapB) {
+      // Total overlap is easy - the new access totally replaces the old.
+      it = writeHistory.erase(it);
+    } else {
+      // The new write partially overlaps a previous write. We want to keep
+      // both, but only track the unconvered part of the earlier write.
+
+      // Determine the slices of the earlier bound not covered by info.
+      auto newBounds =
+          subtractIndicesBounds(indexBounds, info->bounds(), overlap);
+
+      // Erase the old slice.
+      it = writeHistory.erase(it);
+
+      // Add all new slices.
+      for (auto& b : newBounds) {
+        it = writeHistory.insert(it, std::make_pair(b, other));
+      }
+      it++;
+    }
+  }
+
+  if (insert && isWrite) {
+    writeHistory.emplace_back(std::make_pair(info->bounds(), info));
+  }
+}
+
+void MemDependencyChecker::mergeScope(
+    const std::shared_ptr<Scope>& child,
+    const std::shared_ptr<Scope>& parent,
+    bool closeOverlapped) {
+  if (child->accesses_.empty()) {
+    return;
+  }
+
+  int latestAccessToClose =
```

Comment:
It looks like a negative value here is actually meaningful, since `updateWriteHistory` relies on it.

I'd prefer to make this an optional instead and handle the `nullopt` cases separately.

Reply:
This is a bit complex, there are 4 cases here:
1. Adding a new read, which should just have dependencies updated.
1. Adding a new write, which should additionally close open writes in the same scope.
2. Merging a scope into parent scope, and closing open writes in the parent.
3. Merging a scope into a parent scope and NOT closing open writes in the parent (this occurs for conditional scopes that may or may not be entered).

The logic inside updateWriteHistory used to be very complicated and not something I wanted to risk duplicating (it modified a container that it was iterating over using a reverse iterator, which I do not recommend). But perhaps thats not the case now.

Perhaps could split it into 3 functions: Add dependencies, Close overlapping and Add access. We'd loop multiple times but prob not the end of the world.

Reply:
Ah I've remembered the other reason I wanted to roll those 3 together: dependencies and closing overlapping depends on the kind of overlap between the new access and all open accesses in the same scope - I wanted to avoid redoing the work to determine overlaps.

Reply:
I meant to do the `nullopt` handling inside `updateWriteHistory` itself, not here. I don't like the fact that the -1 is involved in comparisons there and we rely on it being smaller than any valid id.

Reply:
I've updated this to separate the two factors (should we close, where to start from) which removes the negative sentinel. What do you think @asuhan ?

Reply:
Yes, this is fine.